### ISSUE TITLE
Convert to WebExtension

### ISSUE THREAD

**Issue by pwd-github** [ddawson/Saved-Password-Editor#28]
As Firefox 57 rapidly approaches given that SPE is a legacy add-on, we're left with the prospect of losing this amazing add-on. To ensure it's future in Firefox, it would need to be converted to a WebExtension.

**Comment by yfdyh000** [comment]
https://bugzilla.mozilla.org/show_bug.cgi?id=1285270
Blocked by https://bugzilla.mozilla.org/show_bug.cgi?id=1324919

**Comment by ddawson** [comment]
Ah, thanks for that, @yfdyh000. I had been thinking I would have to make a request or something myself. Good to know it's in the works.

**Comment by sabret00the** [comment]
From reading the bug, the functionality is available as a [WebExt Experiment](https://github.com/web-ext-experiments/logins) (So can work on Nightly) and will soon be available in Firefox as a standard WebExt API. In short, if SPE was converted to a WebExt now, it would work and will continue to work seamlessly when [Bug 1324919](https://bugzilla.mozilla.org/show_bug.cgi?id=1324919) lands.

**Comment by sabret00the** [comment]
It should be noted that [API access to access and edit](https://bugzilla.mozilla.org/show_bug.cgi?id=1285270) logins has landed and so SPE should be able to exist and work on Nightly. They're doing a security review on its move to release though.

**Comment by ivivanov-bg** [comment]
Any progress on this ?

**Comment by ddawson** [comment]
I woke up this morning and found that Mozilla bug [#1324919](https://bugzilla.mozilla.org/show_bug.cgi?id=1324919) and [#1357856](https://bugzilla.mozilla.org/show_bug.cgi?id=1357856) have had the "wontfix" flag added for Firefox 57. I am not sure if I understand it correctly, but it might mean they're not going to make it in time for that release. Well, I hope that's not it, because if the API doesn't land by then, and they still turn off legacy add-ons, there will be no option to make this add-on continue to work. I should try to find out for sure, though.

**Comment by lnicola** [comment]
Firefox 57 reached beta, so I don't think new features will be added. The API might still land in the future, but not in this version.

**Comment by ivivanov-bg** [comment]
But as far as I understood - the API is already available in Nightly - so may be an unofficial (beta) version of the plugin can be created (and provided for testing for the Nightly users)

**Comment by lnicola** [comment]
If I'm reading it right, the API was available somewhere, but not in the main Firefox codebase.

**Comment by ddawson** [comment]
It's an experimental API. You have to do a custom build of Firefox and manually add the API code in order to test it.

**Comment by ivivanov-bg** [comment]
Or use the already compiled version from here:
https://www.mozilla.org/en-US/firefox/channel/desktop/#nightly

**Comment by lightdot** [comment]
This API being in Nightly would be good news, at least it would arrive sooner rather than later...

BTW, Firefox in the Beta channel just disabled all legacy add-ons, without the possibility of an override that was available in the Nightly.

**Comment by ivivanov-bg** [comment]
>BTW, Firefox in the Beta channel just disabled all legacy add-ons, without the possibility of an override that was available in the Nightly.

How can I do this in Nightly

**Comment by crssi** [comment]
^^
```js
user_pref("extensions.legacy.enabled", true);
```

**Comment by pwd-github** [comment]
> This API being in Nightly would be good news

The API is in Nightly as a WebExtension Experiment. Hopefully @ddawson finds time to release a working WebExtension version for Nightly users until the API makes it way to the releases.

**Comment by AutumnFox** [comment]
Is there any progress on this? It's now November 17, FF 57 is already out and the extension's stopped working :(

**Comment by ddawson** [comment]
@AutumnFox You did read the above comments, right? Especially my first from Sep 26. The "wontfix" tag made it clear (although I was unsure at first) that it's not getting fixed for Fx 57. I still plan to get a WebExtension version out, but it won't help anyone using Fx 57. Nothing I can do to change that fact; that's on the Mozilla devs. Once I do release a WE, if the API still isn't in Firefox, it will be necessary to install the Logins API in its "experiment" form, which I only recommend for developers and those wishing to test, not for regular users. If you want to continue using this right now, there are a few options for you that I know of:

* Downgrade back to Fx 56, and turn off auto-updates.
* Switch to something like Pale Moon (which has stated it will not support WebExtensions), and install my Classic Password Editor in it.
* Do without for a little while.

Also, bear in mind that **some of the functionality is what's always been built-in to Mozilla applications such as Firefox**. They don't allow you to add or edit arbitrary entries, but you can still save the ones the browser prompts you for, and you can see and delete them. Accessing "Saved Logins" is a little less convenient, though.

**Comment by sebma** [comment]
@ddawson Thanks for the Pale Moon web browser alternative, I'll try it right away :+1: :)

**Comment by lnicola** [comment]
@sebma You could also use the Firefox ESR release.

**Comment by ddawson** [comment]
@lnicola Well, there is that. But it's version 52. Is that better than installing the regular 56?

**Comment by lnicola** [comment]
I assume it's better supported in the long run wrt. security updates.

**Comment by uniquePWD** [comment]
An ESR gets security updates that prior versions don't. 

Has work begun on the WebExtension version? Or will work not begin until the API lands?

**Comment by DjDiabolik** [comment]
I see and read some comments about the news firefox 57.x.x and this fantastic extension.
The internal password editors....... it's ok but it's no more possible to create an access if you need:
For example i can't create access for 192.168.1.1 (my routers) and when i try to access firefox don't suggest to me to save the access...........

How i can do ?

**Comment by lilydjwg** [comment]
If you know HTML things, you can press F12 and edit the page to actually submit data via the form (not Ajax magics). I've successfully saved many passwords this way (BTW I need to delete the history containing my password since it's a GET).

**Comment by ddawson** [comment]
On 01/09/2018 06:47 PM, David XXX wrote:
> How i can do ?
A couple of possible ways:

  * Switch to Firefox ESR
    <https://www.mozilla.org/en-US/firefox/organizations/> (Extended
    Support Release), which is currently on version 52, and will remain
    that way until version 59 is released. Hopefully, this issue will be
    resolved by then.
  * Downgrade to a version before 57. Then set the options to not
    automatically update, so it won't update you to 57 or later.

Either of these will allow you to continue using Saved Password Editor.
Alternatively, you could switch to Pale Moon <http://www.palemoon.org/>
(which will not be using WebExtensions) and use Classic Password Editor.

**Comment by DjDiabolik** [comment]
@ddawson yeah i know..... but for now i try to found some workaround....

@lilydjwg it's interesting idea.
I thinks i can try to create a combination of password and username for my router.......

You can post a step to step for an example page ?

**Comment by lilydjwg** [comment]
> You can post a step to step for an example page ?

Sorry, I don't have that at hand and don't have the time to do that recently.

**Comment by DjDiabolik** [comment]
OMG!!

https://superuser.com/questions/781639/manually-add-username-password-to-saved-password-list

IT's works....

Use the HTML for add a custom fied......put the username and password and click Submit.
After that close console and the combination of password and username has been stored inside stock manager!

Now you can use the right context menu for autofill using the admin...........

**Comment by xlollomanx** [comment]
Since FF58 has been released I'm curious if they have implemented the necessary API to make Password editor work again on firefox. Anybody know anything?

**Comment by lnicola** [comment]
@xlollomanx See the [Firefox issue](https://bugzilla.mozilla.org/show_bug.cgi?id=1324919). It's unassigned, so nobody seems to be working on it.

**Comment by DjDiabolik** [comment]
@lnicola this issue it's one years old...... but on last reply we can read this:
All I can see is that this API is not going to get into FF58

No API it's been implement also in FF58 and problably never append...... this project i thinks it's definilly dead.

**Comment by lnicola** [comment]
@DjDiabolik Well, there's probably no interest for Mozilla to add this feature. They have, of course, different priorities than some of the users.

But please don't say anything on that Bugzilla thread, since it never helps and the Mozilla moderators are quite happy to lock issues for "advocacy" posts. There is a "vote" button, but nobody cares about those either.

**Comment by Overc1ocker** [comment]
Although it is unassigned, it seems as though it has caught some of the developers interest

**Comment by Overc1ocker** [comment]
Refer to Djdiabolik. He may have a very simple answer

**Comment by LNx2** [comment]
There is also the possibility to edit the JSON file "logins.json" directly, where the login information is saved. The format is pretty self explanatory and you can just use any of the other entries as examples.

Just increment the "id" counter of the entry, as well as the "nextId" counter at the beginning of the file.
The timestamps are milliseconds since the UNIX epoch, i.e. the first 14 digits of the output of the Unix command "date +%s%N" .
The "guid" entry is a random UUID, i.e. the output of the Unix command "uuidgen --random".

The username and password aren't in clear text, but those can easily be changed through firefox once the entry is created.

As an aside, if you want to use vim for the editing and prefer to have the JSON in a more human-editable format you can use this command in vim:
 `:%!python -c "import json, sys, collections; print json.dumps(json.load(sys.stdin, object_pairs_hook=collections.OrderedDict), indent=4)"` 
from
[https://coderwall.com/p/faceag/format-json-in-vim](url)

Firefox doesn't seem to care about the added whitespaces and newline characters, so you don't even need to strip them back out.

**Comment by jerone** [comment]
I tried [PasswordFox from NirSoft](https://www.nirsoft.net/utils/passwordfox.html) to access all logins. But it doesn't show the real passwords and usernames (because those are encrypted) and it doesn't allow me to alter any of the fields.

-------

Some tutorial to alter the `logins.json` manually: https://www.trishtech.com/2016/12/manually-add-websites-to-password-saving-exception-list-in-firefox/

**Comment by lilydjwg** [comment]
If you want to decrypt those encrypted fields, you can use nss to do that. I have [a simple Python module](https://github.com/lilydjwg/winterpy/blob/master/pylib/firefoxutils.py) to help with that.

**Comment by strawbeard** [comment]
_PasswordFox_ **can decrypt** passwords when you specify the master password: Menu _File | Select Folders_ shows a dialog where you can enter it. It's true that PasswordFox **cannot edit** the password data base. Besides viewing it can do exports to a text, HTML, or CSV (esp. suited for KeePass Password Manager) file.

**Comment by pasko-zh** [comment]
Since [Firefox ESR 52.9 time](https://www.mozilla.org/en-US/firefox/organizations/) is running out in a couple of weeks, I dare to ask if there are any chances to have this _absolutely fine add-on working with webextensions? 
I've read the comments above and I tried to understand/follow the bugs mentioned above... but what is exactly missing in the current and/or updcoming FF webextensions, is it this [login API extension](https://github.com/web-ext-experiments/logins)? Or...? (please excuse my sort of "stupid" asking, but I am more the IoT guy than the webextenstions guy ;)

**Comment by ddawson** [comment]
On 07/18/2018 02:02 AM, paško wrote:
>
> Since Firefox ESR 52.9 time
> <https://www.mozilla.org/en-US/firefox/organizations/> is running out
> in a couple of weeks, I dare to ask if there are any chances to have
> this _absolutely fine add-on working with webextensions?
>
Not in the near future, I'm afraid.

> I've read the comments above and I tried to understand/follow the bugs
> mentioned above... but what is exactly missing in the current and/or
> updcoming FF webextensions, is it this login API extension
> <https://github.com/web-ext-experiments/logins>?
>
No, not that. Probably. Someone started writing a different API later
on, which is the subject of the requested security review.

Actually, what really seems to be missing is any motivation for the
Mozilla devs to move the thing forward, as evidenced by the long gaps of
time in the bug discussions, the lack of good communication, and the
security review bug being closed twice, first as "incomplete" and now as
"invalid". It's been obvious to me for a while they weren't going to get
it done by the time the ESR upgrades, but at this point, I have my
doubts it's ever going to happen.

**Comment by pasko-zh** [comment]
Too bad 😢

**Comment by lnicola** [comment]
Mozilla are working on [their own](https://medium.com/firefox-test-pilot/take-your-passwords-everywhere-with-firefox-lockbox-eee381c4d5af) password management utility, so the Logins API might get ignored or dropped.

**Comment by tomhughes** [comment]
That's just a way to access the existing firefox sync password data from iOS so I don't see what it has to do with this at all...

**Comment by lnicola** [comment]
It's a new password manager using the Firefox Sync infrastructure. If they're planning to recommend that instead of the integrated password manager, I don't think they'll spend resources improving the latter.

**Comment by tomhughes** [comment]
But it's only for iOS devices!

Because everywhere else you can just install firefox and use that to access the synced password store.

**Comment by lnicola** [comment]
> Because everywhere else you can just install firefox and use that to access the synced password store.

I didn't realize that didn't already work on iOS. Anyway, they're also [planning](https://lockbox.firefox.com/#android-mobile-app) an Android version. Not sure about desktop.

**Comment by gerdneuman** [comment]
Have a look at https://lockbox.firefox.com Looks like it's a complete replacement for the existing password manager what would explain why mozilla is not investing time and money into adding functions to the old one. The page has a link to FAQ and Kanban board.

**Comment by tomhughes** [comment]
Well it was, but the web extension part for browser access seems to be have been deprecated recently.

**Comment by pasko-zh** [comment]
@lnicola : Thanks for that, I was not aware of it! The screenshots of the [desktop extension](https://mozilla-lockbox.github.io/lockbox-extension/) look not too bad. Let's see if and when they (Mozilla) add the ability to enter username and password field names manually (one thing I love about saved password editor)

**Comment by UnitedMarsupials-zz** [comment]
Things continue to work for me with Waterfox (56.2.2), but not in Thunderbird (60.0)... Any hope there?

**Comment by lnicola** [comment]
@UnitedMarsupials Read the last comments. Nothing has changed in the meanwhile, as far as I know.

**Comment by NobleUplift** [comment]
I would just like to say thank you to @ddawson for developing this addon that I used for so many years! I still get some websites whose login forms don't trigger a password prompt in Firefox >:(

I have a permanent copy of Firefox 52 installed on my old Windows 7 box that I use for all my pre-Quantum plugins whose functionality is missing in modern Firefox, so I still use this plugin albeit minimally. A sad end to something I've loved using since I was in college.

**Comment by ddawson** [comment]
>A sad end to something I've loved using since I was in college.

An end on Firefox, at least. Not on some other applications. Too bad Firefox is where it has gotten and would continue to get the most use. So, still sad.

**Comment by ddawson** [comment]
@jerone I messed up the quoting on that. Fixed now.

**Comment by uniquePWD** [comment]
Hi everyone, while Mozilla hasn't made the API available to enable the rebirth of this extension in a post-57 world, they have launched Lockwise that contains some of the functionality that this extension held. I would suggest that you head over to that extension's github to file bugs for any functionality you feel is missing and if possible submit pull requests.

You can find the addon here: https://github.com/mozilla-lockwise/lockwise-addon
And the Android version here: https://github.com/mozilla-lockwise/lockwise-android

**Comment by NobleUplift** [comment]
Thanks for the link!!! It feels so good just to have a button on my toolbar that I can use to access my passwords again. I was deeply missing this plugin a few weeks ago when I needed to save a MySQL password but couldn't!

**Comment by NobleUplift** [comment]
Update: You still can't save MySQL passwords in Mozilla Lockwise 😬 https://github.com/mozilla-lockwise/lockwise-addon/issues/328

Also, if you install this extension, it completely removes the Master Password prompt for your passwords, allowing them to be seen in plain text at all times if you leave your computer unlocked. https://github.com/mozilla-lockwise/lockwise-addon/issues/324
